File: Problems/0506_Relative_Ranks/Relative_Ranks.py
# coding: utf-8

import os
import sys
import time
import numpy as np

class Solution:
    def findRelativeRanks(self, nums):
        """
        :type nums: List[int]
        :rtype: List[str]
        """
        if len(nums) == 0: return []
        if len(nums) == 1: return ["Gold Medal"]
        if len(nums) == 2:
            if nums[0] > nums[1]: return ["Gold Medal", "Silver Medal"]
            else: return ["Silver Medal", "Gold Medal"]
        num1 = sorted(nums)[::-1]
        dic = {num: str(i+1) for i, num in enumerate(num1)}
        dic.update({num1[0]:"Gold Medal", num1[1]:"Silver Medal", num1[2]: "Bronze Medal"})
        return [dic[i] for i in nums]

    def findRelativeRanks2(self, nums):
        """
        :type nums: List[int]
        :rtype: List[str]
        """
        nums_ = np.array(nums)
        nums_sort = np.argsort(-nums_)
        
        nums[nums_sort[0]] = "Gold Medal"
        
        if len(nums) >= 2:
            nums[nums_sort[1]] = "Silver Medal"

        if len(nums) >= 3:
            nums[nums_sort[2]] = "Bronze Medal"
        
        if len(nums) > 3:
            for i in range(3, len(nums)):
                nums[nums_sort[i]] = str(i+1)
        
        return nums

    def findRelativeRanks_work(self, nums):
        """
        :type nums: List[int]
        :rtype: List[str]
        """
        if len(nums) <= 0:
            return []
        '''
        sorted_nums = nums[:]
        sorted_nums.sort(reverse = True)
        '''
        sorted_nums = sorted(nums, reverse = True)
        results = [""]*len(nums)
        for n in range(len(results)):
            for i in range(len(sorted_nums)):
                if nums[n] == sorted_nums[i]:
                    if i == 0:
                        results[n] = "Gold Medal"
                    elif i == 1:
                        results[n] = "Silver Medal"
                    elif i == 2:
                        results[n] = "Bronze Medal"
                    else:
                        results[n] = str(i + 1)
        return results

def str_to_int_array(flds):
    nums = [0]*len(flds)
    for i in range(len(flds)):
        nums[i] = int(flds[i])
    return nums

def main():
    argv = sys.argv
    argc = len(argv)

    if (argc < 2):
        print("Usage: python %s <testdata.txt>" %(argv[0]))
        exit(0)

    if not os.path.exists(argv[1]):
        print("%s not found..." %argv[1])
        exit(0)

    testDataFile = open(argv[1], "r")
    lines = testDataFile.readlines()

    for temp in lines:
        print("argv[1] = %s" %temp)
        loop_main(temp)
    #    print("Hit Return to continue...")
    #    input()

def loop_main(temp):
    flds = temp.replace("[","").replace("]","").rstrip().split(",")
    nums = str_to_int_array(flds)

    time0 = time.time()

    sl = Solution()
    result = sl.findRelativeRanks(nums)

    print("result = %s" %result)

    time1 = time.time()
    print("Execute time ... : %f[s]" %(time1 - time0))
    print()

if __name__ == "__main__":
    main()

File: Problems/0500_0599/0506_Relative_Ranks/Project_CS/Relative_Ranks.cs
using System;
using System.Linq;

public class Solution
{
    public string[] FindRelativeRanks(int[] nums)
    {
        string[] medals = new string[] {"Gold Medal", "Silver Medal", "Bronze Medal"};
        // create a sorted decending array and rank them.
        int[] ranks = new int[nums.Length];
        Array.Copy(nums,ranks, nums.Length);
        Array.Sort<int>(ranks, new Comparison<int>((i1, i2) => i2.CompareTo(i1)));
        
        string[] result = new string[nums.Length];
        for(int i = 0; i < nums.Length; ++i){
            int rank = Array.IndexOf(ranks, nums[i]);
            
            if(rank < 3)
                result[i] = medals[rank]; // return rank medal if < 3
            else    
                result[i] = (rank+1).ToString();    // return rank + 1 because index start with 0
        }
        return result;
    }

    public string[] FindRelativeRanks2(int[] nums)
    {
        string[] results = new string[nums.Length];

        if (nums.Length <= 0)
            return results;

        int[] sorted_nums = new int[nums.Length];
        Array.Copy(nums, sorted_nums, nums.Length);
        Array.Sort(sorted_nums);
        Array.Reverse(sorted_nums);

        for (int n = 0; n < results.Length; ++n)
        {
            for (int i = 0; i < sorted_nums.Length; ++i)
            {
                if (nums[n] == sorted_nums[i])
                    if (i == 0)
                        results[n] = "Gold Medal";
                    else if (i == 1)
                        results[n] = "Silver Medal";
                    else if (i == 2)
                        results[n] = "Bronze Medal";
                    else
                        results[n] = (i + 1).ToString();
            }
        }
        return results;
    }

    public int[] str_to_int_array(string[] flds)
    {
        if (flds.Length <= 0)
            return null;

        int[] nums = new int[flds.Length];
        for (int i = 0; i < nums.Length; ++i)
        {
            nums[i] = int.Parse(flds[i]);
        }

        return nums;
    }

    public string output_int_array(int[] flds)
    {
        string results = "";
        if (flds.Length <= 0)
            return results;
        
        results = flds[0].ToString();
        for (int i = 1; i < flds.Length; ++i)
        {
            results += ", " + flds[i].ToString();
        }

        return results;
    }

    public string output_str_array(string[] flds)
    {
        string results = "";
        if (flds.Length <= 0)
            return results;
        
        results = flds[0];
        for (int i = 1; i < flds.Length; ++i)
        {
            results += ", " + flds[i];
        }

        return results;
    }

    public void Main(string args)
    {
        string[] flds = args.Replace("[","").Replace("]","").Trim().Split(',');

        int[] nums = str_to_int_array(flds);
        Console.WriteLine("nums = " + output_int_array(nums));

        System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
        sw.Start();

        string[] results = FindRelativeRanks(nums);
        Console.WriteLine("result = " + output_str_array(results));
        
        sw.Stop();
        Console.WriteLine("Execute time ... " + sw.ElapsedMilliseconds.ToString() + "ms\n");
    }
}
